File: ScaleformUI_Csharp/Menus/Pause Menus/PauseMenu/Tabs/GalleryTab.cs
using ScaleformUI.Elements;
using ScaleformUI.Menus.Pause_Menus.Elements.Columns;
using ScaleformUI.PauseMenus.Elements.Items;
using ScaleformUI.PauseMenus.Elements.Panels;
using System.Reflection;
using static CitizenFX.Core.Native.API;

namespace ScaleformUI.PauseMenu
{
    public enum GalleryState
    {
        EMPTY = 0,
        CORRUPTED = 1,
        QUEUED = 2,
        LOADING = 3,
        LOADED = 4,
        TRANSITION = 5
    }

    public delegate void GalleryModeChanged(GalleryTab tab, GalleryItem item, bool bigPicture);
    public delegate void GalleryIndexChanged(GalleryTab tab, GalleryItem item, int totalIndex, int gridIndex);
    public delegate void GalleryItemSelected(GalleryTab tab, GalleryItem item, int totalIndex, int gridIndex);

    public class GalleryTab : BaseTab
    {
        public event GalleryModeChanged OnGalleryModeChanged;
        public event GalleryIndexChanged OnGalleryIndexChanged;
        public event GalleryItemSelected OnGalleryItemSelected;
        private GalleryColumn column;
        internal bool bigPic = false;

        public GalleryTab(string name, SColor color) : base(name, color)
        {
            _type = 3;
            _identifier = "Page_Gallery";
            column = new GalleryColumn();
            column.Parent = this;
            LeftColumn = column;
            Minimap = new MinimapPanel(this) { HidePedBlip = false };
        }

        public override void Populate()
        {
            column.Index = 0;
            column.currentPageIndex = 0;
            column.Populate();
        }

        public override void GoUp()
        {
            if (!Focused) return;
            LeftColumn.GoUp();
            column.UpdatePage();
            GalleryItem it = column.Items[column.currentPageIndex] as GalleryItem;
            if (bigPic)
            {
                SetTitle(it.TextureDictionary, it.TextureName, GalleryState.LOADED);
            }

            if (it.Blip != null)
            {
                Main.PauseMenu._pause.CallFunction("SET_COLUMN_STATE", (int)column.position, "", false);
                Minimap.Enabled = true;
                Minimap.RefreshMapPosition(new Vector2(it.Blip.Position.X, it.Blip.Position.Y));
            }
            else if (!string.IsNullOrEmpty(it.RightPanelDescription.Label))
            {
                Minimap.Enabled = false;
                var labels = it.RightPanelDescription.SplitLabel;
                BeginScaleformMovieMethod(Main.PauseMenu._pause.Handle, "SET_COLUMN_STATE");
                ScaleformMovieMethodAddParamInt((int)column.position);
                BeginTextCommandScaleformString("CELL_EMAIL_BCON");
                for (var i = 0; i < labels?.Length; i++)
                    AddTextComponentScaleform(labels[i]);
                EndTextCommandScaleformString_2();
                ScaleformMovieMethodAddParamBool(true);
                EndScaleformMovieMethod();
            }
            else
            {
                Minimap.Enabled = false;
                Main.PauseMenu._pause.CallFunction("SET_COLUMN_STATE", (int)column.position, "", true);
            }
        }

        public override void GoDown()
        {
            if (!Focused) return;
            LeftColumn.GoDown();
            column.UpdatePage();
            GalleryItem it = column.Items[column.currentPageIndex] as GalleryItem;
            if (bigPic)
            {
                SetTitle(it.TextureDictionary, it.TextureName, GalleryState.LOADED);
            }

            if (it.Blip != null)
            {
                Main.PauseMenu._pause.CallFunction("SET_COLUMN_STATE", (int)column.position, "", false);
                Minimap.Enabled = true;
                Minimap.RefreshMapPosition(new Vector2(it.Blip.Position.X, it.Blip.Position.Y));
            }
            else if (!string.IsNullOrEmpty(it.RightPanelDescription.Label))
            {
                Minimap.Enabled = false;
                var labels = it.RightPanelDescription.SplitLabel;
                BeginScaleformMovieMethod(Main.PauseMenu._pause.Handle, "SET_COLUMN_STATE");
                ScaleformMovieMethodAddParamInt((int)column.position);
                BeginTextCommandScaleformString("CELL_EMAIL_BCON");
                for (var i = 0; i < labels?.Length; i++)
                    AddTextComponentScaleform(labels[i]);
                EndTextCommandScaleformString_2();
                ScaleformMovieMethodAddParamBool(true);
                EndScaleformMovieMethod();
            }
            else
            {
                Minimap.Enabled = false;
                Main.PauseMenu._pause.CallFunction("SET_COLUMN_STATE", (int)column.position, "", true);
            }
        }

        public override void GoLeft()
        {
            if (!Focused) return;
            LeftColumn.GoLeft();
            column.UpdatePage();
            GalleryItem it = column.Items[column.currentPageIndex] as GalleryItem;
            if (bigPic)
            {
                SetTitle(it.TextureDictionary, it.TextureName, GalleryState.LOADED);
            }

            if (it.Blip != null)
            {
                Main.PauseMenu._pause.CallFunction("SET_COLUMN_STATE", (int)column.position, "", false);
                Minimap.Enabled = true;
                Minimap.RefreshMapPosition(new Vector2(it.Blip.Position.X, it.Blip.Position.Y));
            }
            else if (!string.IsNullOrEmpty(it.RightPanelDescription.Label))
            {
                Minimap.Enabled = false;
                var labels = it.RightPanelDescription.SplitLabel;
                BeginScaleformMovieMethod(Main.PauseMenu._pause.Handle, "SET_COLUMN_STATE");
                ScaleformMovieMethodAddParamInt((int)column.position);
                BeginTextCommandScaleformString("CELL_EMAIL_BCON");
                for (var i = 0; i < labels?.Length; i++)
                    AddTextComponentScaleform(labels[i]);
                EndTextCommandScaleformString_2();
                ScaleformMovieMethodAddParamBool(true);
                EndScaleformMovieMethod();
            }
            else
            {
                Minimap.Enabled = false;
                Main.PauseMenu._pause.CallFunction("SET_COLUMN_STATE", (int)column.position, "", true);
            }
        }

        public override void GoRight()
        {
            if (!Focused) return;
            LeftColumn.GoRight();
            column.UpdatePage();
            GalleryItem it = column.Items[column.currentPageIndex] as GalleryItem;
            if (bigPic)
            {
                SetTitle(it.TextureDictionary, it.TextureName, GalleryState.LOADED);
            }

            if (it.Blip != null)
            {
                Main.PauseMenu._pause.CallFunction("SET_COLUMN_STATE", (int)column.position, "", false);
                Minimap.Enabled = true;
                Minimap.RefreshMapPosition(new Vector2(it.Blip.Position.X, it.Blip.Position.Y));
            }
            else if (!string.IsNullOrEmpty(it.RightPanelDescription.Label))
            {
                Minimap.Enabled = false;
                var labels = it.RightPanelDescription.SplitLabel;
                BeginScaleformMovieMethod(Main.PauseMenu._pause.Handle, "SET_COLUMN_STATE");
                ScaleformMovieMethodAddParamInt((int)column.position);
                BeginTextCommandScaleformString("CELL_EMAIL_BCON");
                for (var i = 0; i < labels?.Length; i++)
                    AddTextComponentScaleform(labels[i]);
                EndTextCommandScaleformString_2();
                ScaleformMovieMethodAddParamBool(true);
                EndScaleformMovieMethod();
            }
            else
            {
                Minimap.Enabled = false;
                Main.PauseMenu._pause.CallFunction("SET_COLUMN_STATE", (int)column.position, "", true);
            }
        }

        public override void GoBack()
        {
            if (!Focused) return;
            LeftColumn.GoBack();
            CurrentColumnIndex = 0;
        }

        public override void Select()
        {
            if (!Focused) return;
            CurrentColumnIndex = 1;
            LeftColumn.Select();
        }

        public override void ShowColumns()
        {
            LeftColumn.ShowColumn();
        }

        public override void Focus()
        {
            base.Focus();
            if ((LeftColumn.Items[LeftColumn.Index] as GalleryItem).Blip != null)
            {
                Main.PauseMenu._pause.CallFunction("SET_COLUMN_STATE", (int)LeftColumn.position, "", false);
                Minimap.Enabled = true;
                Minimap.RefreshMapPosition(new Vector2((LeftColumn.Items[LeftColumn.Index] as GalleryItem).Blip.Position.X, (LeftColumn.Items[LeftColumn.Index] as GalleryItem).Blip.Position.Y));
            }
            else if (!string.IsNullOrEmpty((LeftColumn.Items[LeftColumn.Index] as GalleryItem).RightPanelDescription.Label))
            {
                Minimap.Enabled = false;
                var labels = (LeftColumn.Items[LeftColumn.Index] as GalleryItem).RightPanelDescription.SplitLabel;
                BeginScaleformMovieMethod(Main.PauseMenu._pause.Handle, "SET_COLUMN_STATE");
                ScaleformMovieMethodAddParamInt((int)LeftColumn.position);
                BeginTextCommandScaleformString("CELL_EMAIL_BCON");
                for (var i = 0; i < labels?.Length; i++)
                    AddTextComponentScaleform(labels[i]);
                EndTextCommandScaleformString_2();
                ScaleformMovieMethodAddParamBool(true);
                EndScaleformMovieMethod();
            }
            else
            {
                Minimap.Enabled = false;
                Main.PauseMenu._pause.CallFunction("SET_COLUMN_STATE", (int)LeftColumn.position, "", true);
            }
        }

        public override void UnFocus()
        {
            base.UnFocus();
            Minimap.Enabled = false;
            Main.PauseMenu._pause.CallFunction("SET_COLUMN_STATE", (int)LeftColumn.position, "", true);
        }


        internal void SetTitle(string txd, string txn, GalleryState state)
        {
            column.SetTitle(txd, txn, state);
        }

        public void SetDescriptionLabels(string title, string date, string location, string track, bool visible)
        {
            column.titleLabel = title;
            column.dateLabel = date;
            column.locationLabel = location;
            column.trackLabel = track;
            column.labelsVisible = visible;
            Main.PauseMenu._pause.CallFunction("SET_DESCRIPTION", (int)column.position, LeftColumn.VisibleItems, title, date, location, track, visible);
        }

        public void AddItem(GalleryItem item)
        {
            LeftColumn.AddItem(item);
            if (item.Blip != null)
                Minimap.MinimapBlips.Add(item.Blip);
            if (Parent != null && Parent.Visible && Visible)
            {
                if (column.Items.Count < 12)
                {
                    column.Populate();
                }
            }
        }

        public override void MouseEvent(int eventType, int context, int index)
        {
            if(Focused) return;
            switch (eventType)
            {
                case 5:
                    if(index != column.Index)
                    {
                        column.Items[column.Index].Selected = false;
                        column.Index = index;
                        column.Items[column.Index].Selected = true;
                        break;
                    }
                    column.Select();
                    CurrentColumnIndex = 1;
                    break;
                case 8:
                    break;
                case 9:
                    break;
                case 10:
                case 11:
                    column.MouseScroll(eventType == 10 ? -1 : 1);
                    break;
            }
        }


        internal void ModeChanged()
        {
            OnGalleryModeChanged?.Invoke(this, (column.Items[column.Index] as GalleryItem), bigPic);
        }

        internal void IndexChanged()
        {
            OnGalleryIndexChanged?.Invoke(this, (column.Items[column.Index] as GalleryItem), column.Index, column.currentPageIndex);
        }

        internal void ItemSelected()
        {
            OnGalleryItemSelected?.Invoke(this, (column.Items[column.Index] as GalleryItem), column.Index, column.currentPageIndex);
        }
    }
}

File: ScaleformUI_Lua/src/Menus/PauseMenus/PauseMenu/Tabs/GalleryTab.lua
GalleryTab = {}
GalleryTab.__index = GalleryTab
setmetatable(GalleryTab, { __index = BaseTab })
GalleryTab.__call = function() return "GalleryTab" end

---@class GalleryTab
---@field public Base BaseTab
---@field public LeftItemList PauseMenuItem[]
---@field public Label string
---@field public TextTitle string
---@field public SettingsColumn SettingsListColumn
---@field public PlayersColumn PlayerListColumn
---@field public MissionsColumn MissionListColumn
---@field public StoreColumn StoreListColumn
---@field public Index number
---@field public Focused boolean
---@field public GalleryItems GalleryItem[]
---@field private txd string
---@field private txn string
---@field private maxItemsPerPage number
---@field private titleLabel string
---@field private dateLabel string
---@field private locationLabel string
---@field private trackLabel string
---@field private labelsVisible boolean
---@field private state number
---@field private bigPic boolean
---@field private CurPage number
---@field private currentSelection number
---@field private currentIndex number

---Creates a new Gallery tab.
---@param name string
---@param color SColor
---@param newStyle boolean
---@return GalleryTab
function GalleryTab.New(name, color)
    local base = BaseTab.New(name, color)
    base._identifier = "Page_Gallery"
    base.LeftColumn = GalleryColumn.New()
    base.bigPic = false
    base.OnFocusChanged = function(focus)
    end
    base.OnGalleryModeChanged = function(tab, item, bigPicture)
    end
    base.OnGalleryIndexChanged = function(tab, item, totalIndex, gridIndex)
    end
    base.OnGalleryItemSelected = function(tab, item, totalIndex, gridIndex)
    end
local meta = setmetatable(base, GalleryTab)
    meta.LeftColumn.Parent = meta
    meta.Minimap = MinimapPanel.New(meta)
    return meta
end

function GalleryTab:Populate()
    self.LeftColumn:Index(1)
    self.LeftColumn.currentPageIndex = 1
    self.LeftColumn:Populate()
end

function GalleryTab:GoUp()
    if not self.Focused then return end
    self.LeftColumn:GoUp()
    self.LeftColumn:UpdatePage()
    local it = self.LeftColumn.Items[self.LeftColumn.currentPageIndex]
    if self.bigPic then
        self:SetTitle(it.TextureDictionary, it.TextureName, 4)
    end

    if it.Blip ~= nil then
        ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_COLUMN_STATE", self.LeftColumn.position, "", false)
        self.Minimap:Enabled(true)
        self.Minimap:RefreshMapPosition(vector2(it.Blip.Position.x, it.Blip.Position.y))
    elseif not it.RightPanelDescription:IsNullOrEmpty() then
        self.Minimap:Enabled(false)
        local labels = it.RightPanelDescription:SplitLabel()
        BeginScaleformMovieMethod(ScaleformUI.Scaleforms._pauseMenu._pause.handle, "SET_COLUMN_STATE");
        ScaleformMovieMethodAddParamInt(self.LeftColumn.position);
        BeginTextCommandScaleformString("CELL_EMAIL_BCON");
        for i=1, #labels, 1 do
            AddTextComponentScaleform(labels[i]);
        end
        EndTextCommandScaleformString_2();
        ScaleformMovieMethodAddParamBool(true);
        EndScaleformMovieMethod();
    else
        self.Minimap:Enabled(false)
        ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_COLUMN_STATE", self.LeftColumn.position, "", true)
    end
end

function GalleryTab:GoDown()
    if not self.Focused then return end
    self.LeftColumn:GoDown()
    self.LeftColumn:UpdatePage()
    local it = self.LeftColumn.Items[self.LeftColumn.currentPageIndex]
    if self.bigPic then
        self:SetTitle(it.TextureDictionary, it.TextureName, 4)
    end

    if it.Blip ~= nil then
        ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_COLUMN_STATE", self.LeftColumn.position, "", false)
        self.Minimap:Enabled(true)
        self.Minimap:RefreshMapPosition(vector2(it.Blip.Position.x, it.Blip.Position.y))
    elseif not it.RightPanelDescription:IsNullOrEmpty() then
        self.Minimap:Enabled(false)
        local labels = it.RightPanelDescription:SplitLabel()
        BeginScaleformMovieMethod(ScaleformUI.Scaleforms._pauseMenu._pause.handle, "SET_COLUMN_STATE");
        ScaleformMovieMethodAddParamInt(self.LeftColumn.position);
        BeginTextCommandScaleformString("CELL_EMAIL_BCON");
        for i=1, #labels, 1 do
            AddTextComponentScaleform(labels[i]);
        end
        EndTextCommandScaleformString_2();
        ScaleformMovieMethodAddParamBool(true);
        EndScaleformMovieMethod();
    else
        self.Minimap:Enabled(false)
        ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_COLUMN_STATE", self.LeftColumn.position, "", true)
    end
end

function GalleryTab:GoLeft()
    if not self.Focused then return end
    self.LeftColumn:GoLeft()
    self.LeftColumn:UpdatePage()
    local it = self.LeftColumn.Items[self.LeftColumn.currentPageIndex]
    if self.bigPic then
        self:SetTitle(it.TextureDictionary, it.TextureName, 4)
    end

    if it.Blip ~= nil then
        ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_COLUMN_STATE", self.LeftColumn.position, "", false)
        self.Minimap:Enabled(true)
        self.Minimap:RefreshMapPosition(vector2(it.Blip.Position.x, it.Blip.Position.y))
    elseif not it.RightPanelDescription:IsNullOrEmpty() then
        self.Minimap:Enabled(false)
        local labels = it.RightPanelDescription:SplitLabel()
        BeginScaleformMovieMethod(ScaleformUI.Scaleforms._pauseMenu._pause.handle, "SET_COLUMN_STATE");
        ScaleformMovieMethodAddParamInt(self.LeftColumn.position);
        BeginTextCommandScaleformString("CELL_EMAIL_BCON");
        for i=1, #labels, 1 do
            AddTextComponentScaleform(labels[i]);
        end
        EndTextCommandScaleformString_2();
        ScaleformMovieMethodAddParamBool(true);
        EndScaleformMovieMethod();
    else
        self.Minimap:Enabled(false)
        ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_COLUMN_STATE", self.LeftColumn.position, "", true)
    end
end

function GalleryTab:GoRight()
    if not self.Focused then return end
    self.LeftColumn:GoRight()
    self.LeftColumn:UpdatePage()
    local it = self.LeftColumn.Items[self.LeftColumn.currentPageIndex]
    if self.bigPic then
        self:SetTitle(it.TextureDictionary, it.TextureName, 4)
    end

    if it.Blip ~= nil then
        ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_COLUMN_STATE", self.LeftColumn.position, "", false)
        self.Minimap:Enabled(true)
        self.Minimap:RefreshMapPosition(vector2(it.Blip.Position.x, it.Blip.Position.y))
    elseif not it.RightPanelDescription:IsNullOrEmpty() then
        self.Minimap:Enabled(false)
        local labels = it.RightPanelDescription:SplitLabel()
        BeginScaleformMovieMethod(ScaleformUI.Scaleforms._pauseMenu._pause.handle, "SET_COLUMN_STATE");
        ScaleformMovieMethodAddParamInt(self.LeftColumn.position);
        BeginTextCommandScaleformString("CELL_EMAIL_BCON");
        for i=1, #labels, 1 do
            AddTextComponentScaleform(labels[i]);
        end
        EndTextCommandScaleformString_2();
        ScaleformMovieMethodAddParamBool(true);
        EndScaleformMovieMethod();
    else
        self.Minimap:Enabled(false)
        ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_COLUMN_STATE", self.LeftColumn.position, "", true)
    end
end

function GalleryTab:Select()
    if not self.Focused then return end
    self.CurrentColumnIndex = 1
    self.LeftColumn:Select()
end

function GalleryTab:GoBack()
    if not self.Focused then return end
    self.LeftColumn:GoBack()
    self.CurrentColumnIndex = 0
end

function GalleryTab:ShowColumns()
    self.LeftColumn:ShowColumn()
end

function GalleryTab:Focus()
    BaseTab.Focus(self)
    self.LeftColumn:UpdatePage()
    local it = self.LeftColumn.Items[self.LeftColumn:Index()]
    if it.Blip ~= nil then
        ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_COLUMN_STATE", self.LeftColumn.position, "", false)
        self.Minimap:Enabled(true)
        self.Minimap:RefreshMapPosition(vector2(it.Blip.Position.x, it.Blip.Position.y))
    elseif not it.RightPanelDescription:IsNullOrEmpty() then
        self.Minimap:Enabled(false)
        local labels = it.RightPanelDescription:SplitLabel()
        BeginScaleformMovieMethod(ScaleformUI.Scaleforms._pauseMenu._pause.handle, "SET_COLUMN_STATE");
        ScaleformMovieMethodAddParamInt(self.LeftColumn.position);
        BeginTextCommandScaleformString("CELL_EMAIL_BCON");
        for i=1, #labels, 1 do
            AddTextComponentScaleform(labels[i]);
        end
        EndTextCommandScaleformString_2();
        ScaleformMovieMethodAddParamBool(true);
        EndScaleformMovieMethod();
    else
        self.Minimap:Enabled(false)
        ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_COLUMN_STATE", self.LeftColumn.position, "", true)
    end
end

function GalleryTab:UnFocus()
    BaseTab.UnFocus(self)
    self.Minimap:Enabled(false)
    ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_COLUMN_STATE", self.LeftColumn.position, "", true)
end

function GalleryTab:SetTitle(txd, txn, state)
    self.LeftColumn:SetTitle(txd, txn, state)
end

function GalleryTab:SetDescriptionLabels(title, date, location, track, visible)
    self.LeftColumn.titleLabel = title;
    self.LeftColumn.dateLabel = date;
    self.LeftColumn.locationLabel = location;
    self.LeftColumn.trackLabel = track;
    self.LeftColumn.labelsVisible = visible;
    ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_DESCRIPTION", self.LeftColumn.position, title, date, location, track, visible)
end

function GalleryTab:AddItem(item)
    self.LeftColumn:AddItem(item)
    if item.Blip ~= nil then
        table.insert(self.Minimap.MinimapBlips, item.Blip)
    end
    if self.Parent ~= nil and self.Parent:Visible() and self.Visible then
        if #self.LeftColumn.Items < 12 then
            self.LeftColumn:Populate()
        end
    end
end

function GalleryTab:MouseEvent(eventType, context, index)
    if not self.Focused then return end
    if eventType == 5 then
        if index + 1 ~= self.LeftColumn:Index() then
            self.LeftColumn.Items[self.LeftColumn:Index()]:Selected(false)
            self.LeftColumn:Index(index)
            self.LeftColumn.Items[self.LeftColumn:Index()]:Selected(true)
            return
        end
        self.LeftColumn:Select()
        self.CurrentColumnIndex = 1
    elseif eventType == 10 or eventType == 11 then
        local dir = -1
        if eventType == 11 then
            dir = 1
        end
        self.LeftColumn:MouseScroll(dir)
    end
end